### ISSUE TITLE
Original build system and portability issues

### ISSUE THREAD

**Issue by loganjeverett** [terhorst/EandR-timeseries#1]
Hi, I'm trying to compile this code but none of the machines I have access to have a new enough version of GCC with the required C++11 compliance.  As best I can tell, many Linux distributions still don't even have a compliant version of GCC in their official package tree at all.  So it seems like the C++11 compliance actually places a big barrier to portability of this code, even for vanilla linux builds.  I realize C++11 isn't super new, but the pace at which new GCC versions get adopted into production-grade linux distributions is glacial (not to mention the pace at which most sys admins update OS versions and core packages on shared servers).  So I'm wondering if you can clarify:
1) What parts of the code actually require C++11? Was it used for a few syntax conveniences, or is this code heavily dependent on C++11?
2) What type of system did you originally build this on, and was GCC 4.8 standard or did you install a separate version of GCC specifically for developing this software?
Thank you,
-Logan

**Comment by terhorst** [comment]
The code makes extensive use of memoization to speed up certain recursive function calls, so something like std::map (aka boost::map) is needed. However, I since switched most it over to Google's sparsehash library, so I think it should be fairly easy to remove the C++11 dependency. I'll look into this. Regarding build environment, I use recent versions of Ubuntu (13.10 and 14.04) and OS X (Mavericks and Yosemite). I believe C++11 support is now standard on all of those platforms via either GCC or Clang.

**Comment by loganjeverett** [comment]
Thanks, that is good to know - this looks like a great method and I'm eager to give it a try!  I guess CentOS might be the odd case of still not having C++11 support, but it seems to be a popular choice for university servers in my limited experience, and in any case it's what I'm stuck with right now.  

I managed to install g++ 4.8.4 in my user directory (I also don't have root access, part of the problem on my end, but one that I suspect other people will have as well), and everything seems to compile OK ("make test" works without error), but when I try to run test.py (under python 2.7), I get this error:

`Traceback (most recent call last):`
`File "test.py", line 4, in <module>`
`import pyexp`
`ImportError: libcexp.so: cannot open shared object file: No such file or directory`

libcexp.so is definitely there, does it need to be moved somewhere else on the system?

I tried compiling on OS X Mavericks, but quickly hit a different error:

`clang: error: unsupported option '-static-libgcc'`

I guess this is not a C++11 issue, I'll try just dropping that flag for the OS X build.

Side note: I'm using the version of the code that was included in supporting information for the paper - is that the most current version, or should I be using what's in this git repository?

**Comment by loganjeverett** [comment]
I seem to have gotten past the issue of not finding libcexp.so - basically this file just needs to by moved to a directory like /usr/lib, or if you don't have root access in my case, ~/.local/lib/ and make sure to add this path to LD_LIBRARY_PATH environment variable (and make sure to export it afterwards, that's why it wasn't working for me initially).

When I run bin/test, I now get this output:
`0.000191657`
`0.000400172`
Is this the correct output that indicates the test binary is working?

**Comment by terhorst** [comment]
Use the version that accompanies the paper. (I need to update this repository to match that.) If you are referring to test.cpp in the threelocus/ directory, that's just to check that it compiles and runs. See the file example.py in the directory above it for an interpretable usage example.

**Comment by terhorst** [comment]
Regarding the C++11 issue, the code makes more use of some those features (scoped enums, range-based for loops, std::pair/array/map, to name a few) than I had remembered. It's possible to eliminate all of those but I am pressed for time right now and worry about introducing bugs if I were to rewrite it in a hurry. For now I will leave the C++11 part alone in favor of releasing pre-compiled static binaries for common platforms.

**Comment by loganjeverett** [comment]
Thanks for your help, I was able to get example.py running successfully on CentOS and Mac OS X with a bit of trial and error.  I agree it's not worth removing the C++11 features (and in any case, was not expecting you to do it, I was going to attempt it myself if it was just a matter of swapping in some boost structures).  I realize this is not a real "issue" with your code, but figured this was the right place to discuss it so others could quickly find the solution if they run into the same problem.  What worked for me was the following:

On CentOS server with no root access: Install GCC 4.8 from source in my home directory, and alias the resulting binaries to gcc-4.8, g++-4.8 so that your Makefile uses the right versions.

On Mac OS X 10.9 (Mavericks), I installed homebrew and used that to install true GCC 4.8 (which get installed as gxx-4.8 by default).  Clang does support C++11 but I hit other errors when compiling.

There were a couple other steps I had to do to finally get example.py to see the compiled libraries, but I'll post those in a separate issue.
